### PR TITLE
Init structure

### PR DESCRIPTION
**Pull request by franamor98**
## **Initialization of Repository**

**Changes:**
- Folder structure refactor
- Functions to convert the mesh to torch graphs 
- Function to interpolate element information into nodes [WIP - To be changed]
- Custom torch Datasets to load, fetch and normalize graph information.
- A simple GCNN with 2 GConv layers, pooling and a linear layer. 
- Notebook for visualizations and explain the code done.

### CODE REVIEW

**Comment by franamor98** [comment]
Hello the first PR is here :)

Let me know if there are things to  be updated or correct. Same if you have suggestions/preferences for coding style.
I will work on the ideas discussed by mail.

All the best!
Fran

File: src/utils/graph_utils.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+        if len(features)>0:
+            node_features = torch.cat(features).t().to(torch.float32)
+        else:
+            raise ValueError("None of the features specified is on the graph")
+
+        #find all distinct (undirected) edges from every triangle
+        edges = []
+        for triangle in hour_graph['t']:
+            edges += [ tuple(triangle[:2]), tuple(triangle[1:]), tuple(triangle[[0,-1]]) ] #tuples since we set() it later
```
Comment (akorosov) [inline src/utils/graph_utils.py]: Try using numpy vectorization instead of loops:
```python
edges = np.vstack([
    hour_graph['t'][:2],
    hour_graph['t'][1:],
    hour_graph['t'][[0,-1]],
])
```

File: src/utils/graph_utils.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        # and we stack it as a 2xE (2 for each edge end, E as number of edges)
+        coord_idx= [i for i,key in enumerate(idx_list) if key in ['x','y']]
+        if len(coord_idx)==2:
+            edges_coordinates = [
+                torch.stack(
+                    [
+                        node_features[edge_row][:,coord_idx[0]],
+                        node_features[edge_row][:,coord_idx[1]]
+                    ]
+                )
```
Comment (akorosov) [inline src/utils/graph_utils.py:121]: Maybe vectorisation can be used here as well? I don't know if toch tensor allows that. Numpy does:
`node_features[edges, coord_idx[0]]`

**Review by akorosov — CHANGES_REQUESTED** [review]
Very nice! Great that you use docstrings and created a neat structure from the start!

Did you forget to add __init__.py files to src subfolders?
 
Minor comment / suggestion : try using numpy vectorisation instead of loops. It is usually faster. In my taste it is also easier to understand.

File: src/utils/graph_utils.py (hunk 1 above)
Comment (franamor98) [reply src/utils/graph_utils.py]: Thats right!
I should be careful with the loops. The main reason was that you need to access indexes inside each triangle, and I wanted them to be a list of tuples, so the preserve "identity" to latter cast it into a set a remove repeated edges. I think this would do the same work and faster:

```
edges = np.concatenate([
        hour_graph['t'].transpose()[:2],
        hour_graph['t'].transpose()[1:],
        hour_graph['t'].transpose()[0:-1]
    ],axis=-1)
 edges = torch.tensor(np.unique(edges,axis=-1))

```

File: src/utils/graph_utils.py (hunk 2 above)
Comment (franamor98) [reply src/utils/graph_utils.py:121]: I can not see the way here right now. Torch works as numpy in 99% of the cases :) 
In here we want to index with tuples to retrieve x,y in both ends, if we would use the whole edge tensor (2,n_edges) the indexing wouldn't make sense, I think. There might be another way, but cant see it now, my mind still works in "slow motion" with graphs on tensors. hehe

**Comment by franamor98** [comment]
The init is also added!
I will see some ideas about including the elements in a more elegant way.
Also about a prediction framework. (eg: how big should be the input field, fokker-plank encoding?, 1vsmultiple trajectories...)
